File: src/output/text.ts
// Human-friendly text rendering for the known view models.
import type {
  ComponentDetail,
  DocsResult,
  ListResult,
  QueryResult,
} from "./types.js";

function firstLine(s: string): string {
  const i = s.indexOf("\n");
  return i >= 0 ? s.slice(0, i) : s;
}

function pad(s: string, width: number): string {
  return s.length >= width ? s : s + " ".repeat(width - s.length);
}

function warnings(lines: string[] | undefined): string {
  return (lines ?? []).map((m) => `! ${m}\n`).join("");
}

export function renderQuery(r: QueryResult): string {
  let out = warnings(r.warnings);
  out += `Query: ${r.term}\n`;
  out += `\nComponents (${r.components.length}):\n`;
  for (const c of r.components) {
    out += `  ${pad(c.name, 28)} ${c.id}\n`;
    if (c.description) out += `      ${firstLine(c.description)}\n`;
  }
  out += `\nDocs (${r.docs.length}):\n`;
  for (const d of r.docs) {
    out += `  ${pad(d.title, 28)} ${d.id}\n`;
  }
  return out;
}

export function renderList(r: ListResult): string {
  let out = warnings(r.warnings);
  for (const c of r.components) {
    out += `${pad(c.name, 32)} ${c.id}\n`;
  }
  return out;
}

export function renderDetail(d: ComponentDetail): string {
  let out = warnings(d.warnings);
  out += `${d.name} (${d.id})\n`;
  if (d.description) out += `\n${d.description}\n`;
  if (d.tags && Object.keys(d.tags).length > 0) {
    out += `\nTags:\n`;
    for (const [k, v] of Object.entries(d.tags)) {
      out += `  @${k} ${firstLine(v)}\n`;
    }
  }
  if (d.import) out += `\nImport:\n  ${d.import}\n`;
  if (d.sourceFile) out += `Source: ${d.sourceFile}\n`;
  out += `\nProps (${d.props.length}):\n`;
  for (const p of d.props) {
    const req = p.required ? " (required)" : "";
    const def = p.default ? ` = ${p.default}` : "";
    out += `  ${p.name}: ${p.type}${def}${req}\n`;
    if (p.description) out += `      ${firstLine(p.description)}\n`;
  }
  out += `\nStories (${d.stories.length}):\n`;
  for (const s of d.stories) {
    out += `  ${s.name}\n`;
  }
  if (d.guideline) {
    out += `\nGuideline: ${d.guideline.title}\n\n${d.guideline.content}\n`;
  }
  return out;
}

export function renderDocs(r: DocsResult): string {
  let out = warnings(r.warnings);
  out += `Query: ${r.term}\n`;
  for (const d of r.docs) {
    out += `\n=== ${d.title} (${d.id}) ===\n${d.content}\n`;
  }
  return out;
}

File: internal/output/text.go
package output

import (
	"fmt"
	"io"
	"sort"
	"strings"
)

func sortProps(p []PropDetail) {
	sort.SliceStable(p, func(i, j int) bool { return p[i].Name < p[j].Name })
}

// ew is an error-tracking writer: it records the first write error and turns
// every subsequent Printf into a no-op, so callers can write many lines and
// check err once at the end.
type ew struct {
	w   io.Writer
	err error
}

func (e *ew) printf(format string, args ...any) {
	if e.err != nil {
		return
	}
	_, e.err = fmt.Fprintf(e.w, format, args...)
}

// encodeText renders a human-friendly representation for known view models.
func encodeText(w io.Writer, v any) error {
	e := &ew{w: w}
	switch r := v.(type) {
	case QueryResult:
		writeQuery(e, r)
	case ListResult:
		writeList(e, r)
	case ComponentDetail:
		writeDetail(e, r)
	case DocsResult:
		writeDocs(e, r)
	default:
		// Fall back to JSON for anything without a text renderer.
		return encodeJSON(w, v)
	}
	return e.err
}

func writeWarnings(e *ew, warnings []string) {
	for _, msg := range warnings {
		e.printf("! %s\n", msg)
	}
}

func writeQuery(e *ew, r QueryResult) {
	writeWarnings(e, r.Warnings)
	e.printf("Query: %s\n", r.Term)
	e.printf("\nComponents (%d):\n", len(r.Components))
	for _, c := range r.Components {
		e.printf("  %-28s %s\n", c.Name, c.ID)
		if c.Description != "" {
			e.printf("      %s\n", firstLine(c.Description))
		}
	}
	e.printf("\nDocs (%d):\n", len(r.Docs))
	for _, d := range r.Docs {
		e.printf("  %-28s %s\n", d.Title, d.ID)
	}
}

func writeList(e *ew, r ListResult) {
	writeWarnings(e, r.Warnings)
	for _, c := range r.Components {
		e.printf("%-32s %s\n", c.Name, c.ID)
	}
}

func writeDetail(e *ew, d ComponentDetail) {
	writeWarnings(e, d.Warnings)
	e.printf("%s (%s)\n", d.Name, d.ID)
	if d.Description != "" {
		e.printf("\n%s\n", d.Description)
	}
	if len(d.Tags) > 0 {
		e.printf("\nTags:\n")
		for k, v := range d.Tags {
			e.printf("  @%s %s\n", k, firstLine(v))
		}
	}
	if d.Import != "" {
		e.printf("\nImport:\n  %s\n", d.Import)
	}
	if d.SourceFile != "" {
		e.printf("Source: %s\n", d.SourceFile)
	}
	e.printf("\nProps (%d):\n", len(d.Props))
	for _, p := range d.Props {
		req := ""
		if p.Required {
			req = " (required)"
		}
		def := ""
		if p.Default != "" {
			def = fmt.Sprintf(" = %s", p.Default)
		}
		e.printf("  %s: %s%s%s\n", p.Name, p.Type, def, req)
		if p.Description != "" {
			e.printf("      %s\n", firstLine(p.Description))
		}
	}
	e.printf("\nStories (%d):\n", len(d.Stories))
	for _, s := range d.Stories {
		e.printf("  %s\n", s.Name)
	}
	if d.Guideline != nil {
		e.printf("\nGuideline: %s\n\n%s\n", d.Guideline.Title, d.Guideline.Content)
	}
}

func writeDocs(e *ew, r DocsResult) {
	writeWarnings(e, r.Warnings)
	e.printf("Query: %s\n", r.Term)
	for _, d := range r.Docs {
		e.printf("\n=== %s (%s) ===\n%s\n", d.Title, d.ID, d.Content)
	}
}

func firstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}
